R/arc-read.R

```diff
@@ -68,109 +75,175 @@ arc_read <- function(
     crs = NULL,
     ...,
     fields = NULL,
+    alias = c("drop", "label", "replace"),
     token = arc_token()
 ) {
-  service <- arc_open(url = url, token = token)
+  x <- arc_open(url = url, token = token)
 
-  crs <- crs %||% sf::st_crs(service)
+  # Default crs must be NULL since crs can't be taken from x at execution
+  crs <- crs %||% sf::st_crs(x)
 
-  # if the server is an ImageServer we use arc_raster
-  if (inherits(service, "ImageServer")) {
+  # if the server is an ImageServer use arc_raster
+  if (inherits(x, "ImageServer")) {
     layer <- arc_raster(
-      x = service,
+      x = x,
       ...,
       crs = crs,
       token = token
     )
 
     return(layer)
 
-  } else if (!obj_is_layer(service)) {
+  } else if (!obj_is_layer(x)) {
     # if it is not a layer we abort
     # implicitly checks for Layer type and permits continuing
     cli::cli_abort(
       c(
         "{.arg url} is not a supported type:
       {.val FeatureLayer}, {.val Table}, or {.val ImageServer}",
-        "i" = "found {.val {class(service)}}"
+      "i" = "found {.val {class(x)}}"
       )
     )
   }
 
   layer <- arc_select(
-    x = service,
+    x = x,
     fields = col_select %||% fields,
     crs = crs,
     n_max = n_max,
     token = token,
     ...
   )
 
-  set_layer_names(
+  set_layer_col_names(
     layer,
     col_names = col_names,
     name_repair = name_repair,
-    alias = service[["fields"]][["alias"]]
+    alias = alias,
+    x = x
   )
 }
 
 #' Set names for layer or table
-#'
 #' @noRd
-set_layer_names <- function(
-    x,
-    col_names = NULL,
+set_layer_col_names <- function(
+    layer,
+    col_names = TRUE,
     name_repair = NULL,
-    alias = NULL,
+    alias = c("drop", "label", "replace"),
+    x = NULL,
     call = rlang::caller_env()
 ) {
-  layer_nm <- names(x)
+  alias <- rlang::arg_match(alias, error_call = call)
+  has_name_repair <- !is.null(name_repair)
+
+  if (!rlang::is_logical(col_names, 1) && !is.character(col_names)) {
+    cli::cli_abort(
+      "{.arg col_names} must be `TRUE`, `FALSE`, or a character vector.",
+      call = call
+    )
+  }
+
+  if (rlang::is_true(col_names) && alias == "drop" && !has_name_repair) {
+    return(layer)
+  }
 
   # Use existing names by default
-  nm <- layer_nm
-  sf_column_nm <- attr(x, "sf_column")
+  existing_nm <- names(layer)
```

Comment:
Use colnames just in case something weird happens and row-names exist
```suggestion
  existing_nm <- colnames(layer)
```